src/Funnel.sol

```diff
@@ -198,23 +198,36 @@ contract Funnel is IFunnel, NativeMetaTransaction, MetaTxContext, Initializable
     /// @notice fetch amounts spendable by _spender
     /// @return remaining allowance at the current point in time
     function allowance(address _owner, address _spender)
-        public
+        external
         view
         returns (uint256 remaining)
     {
         return _remainingAllowance(_owner, _spender);
     }
 
+    // @notice fetch remaining allowance between _owner and _spender while accounting for base token allowance.
     function _remainingAllowance(address _owner, address _spender)
         private
         view
         returns (uint256)
     {
         RenewableAllowance memory a = rAllowance[_owner][_spender];
+        uint256 baseAllowance = _baseToken.allowance(_owner, address(this));
+        uint256 recovered = a.recoveryRate * uint64(block.timestamp - a.lastUpdated); // uint192 * uint64 = uint256
+        uint256 remainingAllowance = saturatingAdd(a.remaining, recovered); // uint256 + uint256 potential overflow
+        uint256 currentAllowance = remainingAllowance > a.maxAmount
+            ? a.maxAmount
+            : remainingAllowance;
+        return currentAllowance > baseAllowance ? baseAllowance : currentAllowance;
+    }
 
-        uint256 recovered = a.recoveryRate * (block.timestamp - a.lastUpdated);
-        uint256 remainingAllowance = a.remaining + recovered;
-        return remainingAllowance > a.maxAmount ? a.maxAmount : remainingAllowance;
+    // @dev returns the sum of two uint256 values, saturating at 2**256 - 1
+    function saturatingAdd(uint256 a, uint256 b) internal pure returns (uint256) {
```

Comment:
Where should this function be located? @zhongfu @edison0xyz

Reply:
not sure, maybe some new file in `lib/`? i think `internal` functions end up being inlined anyway, so contract size (and probably gas costs for deploy, by extension?) might stay the same

plus it feels a little weird to have it in `Funnel.sol` since it isn't really funnel-specific

Reply:
Some other codebases will put such helper functions into `lib`. Example: https://github.com/ribbon-finance/ribbon-v2/blob/master/contracts/libraries/ShareMath.sol 

However, I wonder if it is an overkill since we do not have that many functions now.